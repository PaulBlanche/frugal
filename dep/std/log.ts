export * from 'https://deno.land/std@0.167.0/log/mod.ts';
