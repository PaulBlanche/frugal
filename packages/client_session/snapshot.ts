import * as murmur from '../murmur/mod.ts';

type Node = {
    name: string;
    attributes: { name: string; value: string }[];
    content: string;
    hash: string;
    element: Element;
};

export type Head = Node[];

export function serialize(head: HTMLHeadElement) {
    const snapshot: Head = [];

    head.childNodes.forEach((node) => {
        if (!(node instanceof Element)) {
            return;
        }

        const hash = new murmur.Hash();

        const name = node.nodeName.toLowerCase();
        const content = node.textContent ?? '';

        const sorted: string[] = [];
        const attributes: { name: string; value: string }[] = [];
        for (let i = 0, length = node.attributes.length; i < length; i++) {
            const attribute = node.attributes[i];
            attributes.push({
                name: attribute.name,
                value: attribute.value,
            });
            sorted.push(`${attribute.name}="${attribute.value}"`);
        }

        hash.update(name);

        sorted.sort();
        sorted.forEach((value) => {
            hash.update(value);
        });

        hash.update(content);

        snapshot.push({
            name,
            attributes,
            content,
            hash: hash.digest(),
            element: node,
        });
    });

    return snapshot;
}
