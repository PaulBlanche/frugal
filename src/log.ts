import * as colors from '../dep/std/fmt/colors.ts';

const LEVELS = {
    'silent': 0,
    'error': 1,
    'warning': 2,
    'info': 3,
    'debug': 4,
    'verbose': 5,
};

type Level = keyof typeof LEVELS;

type LogEntryConfig = {
    kind: Exclude<Level, 'silent'>;
    scope: string;
    scopeExtra?: string;
    extra: string;
};

export type LogConfig = {
    level: Level;
    scopes: Record<string, Level | undefined>;
};

const GLOBAL_CONFIG: LogConfig = {
    level: 'info',
    scopes: {},
};

export function config(config: Partial<LogConfig>) {
    Object.assign(GLOBAL_CONFIG, config);
}

export type Log = (
    messageOrError: string | Error,
    options?: Partial<
        LogEntryConfig
    >,
) => void;

const TIME_FORMATTER = new Intl.DateTimeFormat('en-US', {
    hour12: false,
    hour: '2-digit',
    minute: '2-digit',
    second: '2-digit',
    fractionalSecondDigits: 3,
    timeZoneName: 'short',
});

const DATE_FOMATTER = new Intl.DateTimeFormat('fr-FR', { dateStyle: 'short' });

export function log(
    messageOrError: string | Error,
    {
        scope = '???',
        scopeExtra,
        kind = messageOrError instanceof Error ? 'error' : 'info',
        extra,
    }: Partial<
        LogEntryConfig
    > = {},
) {
    const level = GLOBAL_CONFIG.scopes[scope] ?? GLOBAL_CONFIG.level;

    if (LEVELS[level] < LEVELS[kind]) {
        return;
    }
    const now = new Date();
    const date = colors.gray(
        `${DATE_FOMATTER.format(now)} ${TIME_FORMATTER.format(now)}`,
    );

    const message = `${date} ${formatKind(kind)} ${formatScope(scope, scopeExtra, kind)} ${
        formatMessage(messageOrError, kind, extra)
    }`;

    console.log(message);
}

function formatMessage(
    messageOrError: Error | string,
    kind: Level,
    extra?: string,
) {
    const msg = [formatMessageContent(messageOrError)];

    const formatedExtra = formatExtra(messageOrError, extra);
    if (formatedExtra) {
        msg.push(formatedExtra);
    }

    const message = msg.join('\n');

    switch (kind) {
        case 'warning':
            return colors.yellow(message);
        case 'error':
            return colors.brightRed(message);
        case 'debug':
        case 'verbose': {
            return colors.gray(message);
        }
        default:
            return message;
    }
}

// deno-lint-ignore no-explicit-any
function formatMessageContent(messageOrError: any) {
    if (messageOrError instanceof Error) {
        return `${messageOrError.name} : ${messageOrError.message}`;
    } else {
        return String(messageOrError);
    }
}

function formatExtra(messageOrError: Error | string, extra?: string) {
    if (!(messageOrError instanceof Error)) {
        return extra;
    }

    return `\n${formatError(messageOrError)}`;
}

// deno-lint-ignore no-explicit-any
function formatCause(cause: any) {
    if (cause instanceof Error) {
        return formatError(cause);
    }
    {
        return String(cause);
    }
}

function formatError(error: Error): string {
    const stack = error.stack ??
        `${error.name} : ${error.message}\n    [no stack]`;

    const msg = [stack];
    if (error.cause) {
        msg.push(`\ncaused by ${formatCause(error.cause)}`);
    }

    return msg.join('\n');
}

function formatScope(
    scope: string,
    scopeExtra: string | undefined,
    kind: Level,
) {
    const fullScope = scopeExtra ? `${scope}:${scopeExtra}` : scope;
    const formattedScope = colors.bold(` ${fullScope} >`);

    switch (kind) {
        case 'warning':
            return colors.yellow(formattedScope);
        case 'error':
            return colors.red(formattedScope);
        case 'debug':
        case 'verbose': {
            return colors.gray(formattedScope);
        }
        default:
            return formattedScope;
    }
}

function formatKind(kind: Level) {
    const content = kind.toUpperCase().padEnd(7);
    switch (kind) {
        case 'debug':
        case 'verbose':
            return cell(content, 8);
        case 'info':
            return cell(content, 15);
        case 'warning':
            return cell(content, 11);
        case 'error':
            return cell(content, 9);
    }
}

function cell(content: string, color: number) {
    const open = colors.rgb8('[', color);
    const close = colors.rgb8(']', color);
    return colors.bgRgb8(
        `${open}${colors.black(content)}${close}`,
        color,
    );
}
