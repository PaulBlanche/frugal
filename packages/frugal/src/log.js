import * as _type from "./_type/log.js";
export * from "./_type/log.js";

import * as colors from "../dependencies/colors.js";

/** @enum {number} */
export const LEVELS = /** @type {const} */ ({
    silent: 0,
    error: 1,
    warning: 2,
    info: 3,
    debug: 4,
    verbose: 5,
});

/** @type {_type.LogConfig} */
const GLOBAL_CONFIG = {
    level: "info",
    scopes: {},
};

/** @param {Partial<_type.LogConfig>} config */
export function config(config) {
    Object.assign(GLOBAL_CONFIG, config);
}

const TIME_FORMATTER = new Intl.DateTimeFormat("en-US", {
    hour12: false,
    timeZone: "UTC",
    hour: "2-digit",
    minute: "2-digit",
    second: "2-digit",
    fractionalSecondDigits: 3,
    timeZoneName: "short",
});

const DATE_FOMATTER = new Intl.DateTimeFormat("fr-FR", { dateStyle: "short" });

/**
 * @param {string | Error} messageOrError
 * @param {Partial<_type.LogEntry>} [param1]
 * @returns {void}
 */
export function log(
    messageOrError,
    { scope = "???", level = messageOrError instanceof Error ? "error" : "info", extra } = {},
) {
    const currentLevel = GLOBAL_CONFIG.scopes[scope] ?? GLOBAL_CONFIG.level;

    if (LEVELS[currentLevel] < LEVELS[level]) {
        return;
    }
    const now = new Date();
    const date = colors.grey(`${DATE_FOMATTER.format(now)} ${TIME_FORMATTER.format(now)}`);

    const message = `${date} ${formatLevel(level)} ${formatScope(scope, level)} ${formatMessage(
        messageOrError,
        level,
        extra,
    )}`;

    console.log(message);
}

/**
 * @param {string | Error} messageOrError
 * @param {_type.Level} level
 * @param {string} [extra]
 * @returns {string}
 */
function formatMessage(messageOrError, level, extra) {
    const msg = [formatMessageContent(messageOrError)];

    const formatedExtra = formatExtra(messageOrError, extra);
    if (formatedExtra) {
        msg.push(formatedExtra);
    }

    const message = msg.join("\n");

    switch (level) {
        case "warning":
            return colors.yellow(message);
        case "error":
            return colors.brightRed(message);
        case "debug":
        case "verbose": {
            return colors.grey(message);
        }
        default:
            return message;
    }
}

/**
 * @param {string | Error} messageOrError
 * @returns {string}
 */
function formatMessageContent(messageOrError) {
    if (messageOrError instanceof Error) {
        return `${messageOrError.name} : ${messageOrError.message}`;
    } else {
        return String(messageOrError);
    }
}

/**
 * @param {string | Error} messageOrError
 * @param {string} [extra]
 * @returns {string | undefined}
 */
function formatExtra(messageOrError, extra) {
    if (!(messageOrError instanceof Error)) {
        return extra;
    }

    return `\n${formatError(messageOrError)}`;
}

/**
 * @param {any} cause
 * @returns {string}
 */
function formatCause(cause) {
    if (cause instanceof Error) {
        return formatError(cause);
    }
    {
        return String(cause);
    }
}

/**
 * @param {Error} error
 * @returns {string}
 */
function formatError(error) {
    const stack = error.stack ?? `${error.name} : ${error.message}\n    [no stack]`;

    const msg = [stack];
    if (error.cause) {
        msg.push(`\ncaused by ${formatCause(error.cause)}`);
    }

    return msg.join("\n");
}

/**
 * @param {string} scope
 * @param {_type.Level} level
 * @returns
 */
function formatScope(scope, level) {
    const formattedScope = colors.bold(` ${scope} >`);

    switch (level) {
        case "warning":
            return colors.yellow(formattedScope);
        case "error":
            return colors.red(formattedScope);
        case "debug":
        case "verbose": {
            return colors.grey(formattedScope);
        }
        default:
            return formattedScope;
    }
}

/**
 * @param {_type.Level} level
 * @returns {string | undefined}
 */
function formatLevel(level) {
    const content = level.toUpperCase().padEnd(7);
    switch (level) {
        case "debug":
        case "verbose":
            return cell(content, colors.grey, colors.bgGray);
        case "info":
            return cell(content, colors.brightWhite, colors.bgBrightWhite);
        case "warning":
            return cell(content, colors.brightYellow, colors.bgBrightYellow);
        case "error":
            return cell(content, colors.brightRed, colors.bgBrightRed);
    }
}

/**
 * @param {string} content
 * @param {(str: string) => string} fg
 * @param {(str: string) => string} bg
 * @returns {string}
 */
function cell(content, fg, bg) {
    return bg(`${fg("[")}${colors.black(content)}${fg("]")}`);
}
