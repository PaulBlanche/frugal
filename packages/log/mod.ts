import * as log from '../../dep/std/log.ts';
import * as colors from '../../dep/std/colors.ts';
import * as murmur from '../murmur/mod.ts';

/**
 * Internal representation of a log record
 */
type FrugalLogRecord = {
    /** the log message */
    msg?: string;
    logger: {
        level: number;
        levelName: string;
        /** the scope of the logger that produced the current record */
        scope: string;
        datetime: Date;
        /** an optional property marking the start of a time interval
         * measurment */
        timerStart?: string;
        /** an optional property marking the end of a time interval measurment
         * (relative to a log record with a matching `startTime`) */
        timerEnd?: string;
        /** a property automatically filled with the result of a time interval
         * measurment (in the `timerEnd` record) */
        delta?: number;
    };
};

/**
 * Small time keeper utility
 */
class StopWatch {
    #timers: Map<string, number>;

    constructor() {
        this.#timers = new Map();
    }

    /**
     * Given a `FrugalLogRecord`, this function will :
     * - start a timer associated with the `logger.timerStart` value (if
     *   present)
     * - end the timer associated with the `logger.timerEnd` value (if present)
     *   and set the `logger.delta` value.
     */
    time(record: FrugalLogRecord) {
        if (record.logger.timerStart) {
            const timerStartKey =
                `${record.logger.scope}:${record.logger.timerStart}`;
            this.#timers.set(timerStartKey, performance.now());
        }

        if (record.logger.timerEnd) {
            const timerEndKey =
                `${record.logger.scope}:${record.logger.timerEnd}`;
            if (this.#timers.has(timerEndKey)) {
                record.logger.delta = performance.now() -
                    this.#timers.get(timerEndKey)!;
            }
        }
    }
}

/**
 * Format a `log.LogRecord` to the `FrugalLogRecord` structure.
 */
class FrugalFormatter {
    #stopwatch: StopWatch;

    constructor() {
        this.#stopwatch = new StopWatch();
    }

    format(logRecord: log.LogRecord): FrugalLogRecord {
        const data = this.#parseData(logRecord);

        const record: FrugalLogRecord = {
            ...data,
            logger: {
                ...data.logger,
                level: logRecord.level,
                levelName: logRecord.levelName,
                scope: logRecord.loggerName,
                datetime: logRecord.datetime,
            },
        };

        this.#stopwatch.time(record);

        return record;
    }

    #parseData(logRecord: log.LogRecord): Partial<FrugalLogRecord> {
        try {
            return JSON.parse(logRecord.msg);
        } catch {
            return { msg: logRecord.msg };
        }
    }
}

/**
 * A basic log handler, that format each log record passing through
 */
class FrugalHandler extends log.handlers.ConsoleHandler {
    #frugalFormatter: FrugalFormatter;

    constructor(levelName: log.LevelName, options: log.HandlerOptions = {}) {
        super(levelName, options);
        this.#frugalFormatter = new FrugalFormatter();
    }

    toFrugalRecord(logRecord: log.LogRecord): FrugalLogRecord {
        return this.#frugalFormatter.format(logRecord);
    }
}

/**
 * The JSONHandler, that flat-out stringify any `FrugalLogRecord` passing
 * through.
 */
class JSONHandler extends FrugalHandler {
    format(logRecord: log.LogRecord): string {
        return JSON.stringify(this.toFrugalRecord(logRecord));
    }
}

/**
 * The JSONHandler, that will output a human readable string for each
 * `FrugalLogRecord` passing through. Messages will be colored depending on
 * levels :
 *  - `CRITICAL` or `ERROR` level messages will be printed black on a red
 *    background
 *  - `WARNING` level messages will be printed yellow
 *  - `INFO` or `DEBUG` level messages will be printed white, with a random
 *    persistent color for each scope.
 *
 * If a `logger.delta` value is present, a string `'(done in xxx s)'` will be
 * appended to the message
 */
class HumanHandler extends FrugalHandler {
    #colors: Map<string, number>;

    constructor(levelName: log.LevelName, options: log.HandlerOptions = {}) {
        super(levelName, options);
        this.#colors = new Map();
    }

    format(logRecord: log.LogRecord): string {
        const record = this.toFrugalRecord(logRecord);

        const time =
            `${record.logger.datetime.toLocaleDateString()} ${record.logger.datetime.toLocaleTimeString()}.${record.logger.datetime.getMilliseconds()}`;
        const scope = record.logger.scope;
        const level = record.logger.levelName;

        switch (level) {
            case 'CRITICAL':
            case 'ERROR': {
                return this.#message(
                    time,
                    level,
                    scope,
                    record,
                );
            }
            case 'WARNING': {
                return colors.yellow(this.#message(time, level, scope, record));
            }
            case 'INFO':
            case 'DEBUG': {
                const coloredScope = colors.rgb8(
                    scope,
                    this.#scopeColorIndex(scope),
                );
                return this.#message(time, level, coloredScope, record);
            }
            default: {
                return this.#message(time, level, scope, record);
            }
        }
    }

    #message(
        time: string,
        level: string,
        scope: string,
        record: FrugalLogRecord,
    ): string {
        const delta = record.logger.delta !== undefined
            ? ` (done in ${record.logger.delta / 1000} s)`
            : '';

        return `[${time}|${level}|${scope}] - ${
            record.msg ?? JSON.stringify(record)
        }${delta}`;
    }

    #scopeColorIndex(scope: string): number {
        if (!this.#colors.has(scope)) {
            const hash = new murmur.Hash().update(scope).digest('raw');

            const index = (hash % Math.floor((6 * 5 * 6 - 3) / 2)) * 2 + 19;

            this.#colors.set(scope, index);
        }

        return this.#colors.get(scope)!;
    }
}

/**
 * The user log record
 */
type FrugalLogEntry = {
    msg?: string | (() => string);
    $$error?: Error;
    logger?: {
        timerStart?: string;
        timerEnd?: string;
        delta?: number;
    };
} & Record<string, unknown>;

/**
 * The logger, wrapping a `log.Logger`. This wrapper is an adapter between the
 * custom `FrugalLogEntry` structure and the `log.Logger`
 */
export class FrugalLogger {
    logger: log.Logger;

    constructor(logger: log.Logger) {
        this.logger = logger;
    }

    #log<T extends FrugalLogEntry>(
        level: 'info' | 'debug' | 'warning' | 'error' | 'critical',
        entry: T,
    ): Omit<T, 'msg'> & { msg?: string } {
        let msg = typeof entry.msg === 'function' ? entry.msg() : entry.msg;
        if (entry.$$error) {
            msg += `\n${entry.$$error.stack}`;
        }
        if (this.logger.level === 0) {
            return { ...entry, msg };
        }

        // deno-lint-ignore no-explicit-any
        return this.logger[level]<T>({ ...entry, msg } as any) as any;
    }

    debug<T extends FrugalLogEntry>(
        entry: T,
    ): Omit<T, 'msg'> & { msg?: string } {
        return this.#log('debug', entry);
    }

    info<T extends FrugalLogEntry>(
        entry: T,
    ): Omit<T, 'msg'> & { msg?: string } {
        return this.#log('info', entry);
    }

    warning<T extends FrugalLogEntry>(
        entry: T,
    ): Omit<T, 'msg'> & { msg?: string } {
        return this.#log('warning', entry);
    }

    error<T extends FrugalLogEntry>(
        entry: T,
        error?: Error,
    ): Omit<T, 'msg'> & { msg?: string } {
        return this.#log('error', {
            ...entry,
            '$$error': error,
        });
    }

    critical<T extends FrugalLogEntry>(
        entry: T,
        error?: Error,
    ): Omit<T, 'msg'> & { msg?: string } {
        return this.#log('critical', {
            ...entry,
            '$$error': error,
        });
    }
}

export type Config = {
    type?: 'human' | 'json';
    loggers: {
        [name: string]: log.LevelName;
    };
};

const DEFAULT_CONFIG: Config = {
    loggers: {
        default: 'DEBUG',
    },
};

export async function setup({ type = 'human', loggers }: Config) {
    const handlers: { [s: string]: HumanHandler | JSONHandler } = {};

    if (type === 'human') {
        handlers['human'] = new HumanHandler('DEBUG');
    } else {
        handlers['json'] = new JSONHandler('DEBUG');
    }

    await log.setup({
        handlers,
        loggers: Object.entries(loggers).reduce((loggers, [name, level]) => {
            loggers[name] = {
                level,
                handlers: [type],
            };
            return loggers;
        }, {} as NonNullable<log.LogConfig['loggers']>),
    });
}

export function getLogger(name?: string) {
    const logger = log.getLogger(name);
    return new FrugalLogger(logger);
}

await setup(DEFAULT_CONFIG);
