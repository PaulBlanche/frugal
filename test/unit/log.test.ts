import { assertSnapshot } from "../../dep/std/testing/snapshot.ts";
import { config, log } from "../../src/log.ts";

Deno.test("log: default config", async (t) => {
    const logs: any[][] = [];
    console.log = (...params) => logs.push(params);

    logTest();

    assertSnapshot(t, logs);
});

Deno.test("log: custom config", async (t) => {
    const logs: any[][] = [];
    console.log = (...params) => logs.push(params);

    config({
        level: "error",
        scopes: {
            test2: "verbose",
        },
    });
    logTest();

    assertSnapshot(t, logs);
});

function logTest() {
    log("default");
    log("error", { scope: "test1", level: "error" });
    log(new Error("true error default"), { scope: "test1" });
    log(new Error("true error"), { scope: "test1", level: "info" });
    log("warning", { scope: "test1", level: "warning" });
    log("info", { scope: "test1", level: "info" });
    log("debug", { scope: "test1", level: "debug" });
    log("verbose", { scope: "test1", level: "verbose" });

    log("error", { scope: "test2", level: "error" });
    log("warning", { scope: "test2", level: "warning" });
    log("info", { scope: "test2", level: "info" });
    log("debug", { scope: "test2", level: "debug" });
    log("verbose", { scope: "test2", level: "verbose" });
}
