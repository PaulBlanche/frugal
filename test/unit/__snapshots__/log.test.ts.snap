export const snapshot = {};

snapshot[`log: default config 1`] = `
[
  [
    "\\x1b[90m01/01/1970 24:00:00.000 UTC\\x1b[39m \\x1b[48;5;15m\\x1b[38;5;15m[\\x1b[39m\\x1b[30mINFO   \\x1b[39m\\x1b[38;5;15m]\\x1b[39m\\x1b[49m \\x1b[1m ??? >\\x1b[22m default",
  ],
  [
    "\\x1b[90m01/01/1970 24:00:02.000 UTC\\x1b[39m \\x1b[48;5;9m\\x1b[38;5;9m[\\x1b[39m\\x1b[30mERROR  \\x1b[39m\\x1b[38;5;9m]\\x1b[39m\\x1b[49m \\x1b[31m\\x1b[1m test1 >\\x1b[22m\\x1b[39m \\x1b[91merror\\x1b[39m",
  ],
  [
    "\\x1b[90m01/01/1970 24:00:04.000 UTC\\x1b[39m \\x1b[48;5;9m\\x1b[38;5;9m[\\x1b[39m\\x1b[30mERROR  \\x1b[39m\\x1b[38;5;9m]\\x1b[39m\\x1b[49m \\x1b[31m\\x1b[1m test1 >\\x1b[22m\\x1b[39m \\x1b[91mError : true error default
" +
      "
" +
      "[mocked error stack]\\x1b[39m",
  ],
  [
    "\\x1b[90m01/01/1970 24:00:06.000 UTC\\x1b[39m \\x1b[48;5;15m\\x1b[38;5;15m[\\x1b[39m\\x1b[30mINFO   \\x1b[39m\\x1b[38;5;15m]\\x1b[39m\\x1b[49m \\x1b[1m test1 >\\x1b[22m Error : true error
" +
      "
" +
      "[mocked error stack]",
  ],
  [
    "\\x1b[90m01/01/1970 24:00:08.000 UTC\\x1b[39m \\x1b[48;5;11m\\x1b[38;5;11m[\\x1b[39m\\x1b[30mWARNING\\x1b[39m\\x1b[38;5;11m]\\x1b[39m\\x1b[49m \\x1b[33m\\x1b[1m test1 >\\x1b[22m\\x1b[39m \\x1b[33mwarning\\x1b[39m",
  ],
  [
    "\\x1b[90m01/01/1970 24:00:10.000 UTC\\x1b[39m \\x1b[48;5;15m\\x1b[38;5;15m[\\x1b[39m\\x1b[30mINFO   \\x1b[39m\\x1b[38;5;15m]\\x1b[39m\\x1b[49m \\x1b[1m test1 >\\x1b[22m info",
  ],
  [
    "\\x1b[90m01/01/1970 24:00:16.000 UTC\\x1b[39m \\x1b[48;5;9m\\x1b[38;5;9m[\\x1b[39m\\x1b[30mERROR  \\x1b[39m\\x1b[38;5;9m]\\x1b[39m\\x1b[49m \\x1b[31m\\x1b[1m test2 >\\x1b[22m\\x1b[39m \\x1b[91merror\\x1b[39m",
  ],
  [
    "\\x1b[90m01/01/1970 24:00:18.000 UTC\\x1b[39m \\x1b[48;5;11m\\x1b[38;5;11m[\\x1b[39m\\x1b[30mWARNING\\x1b[39m\\x1b[38;5;11m]\\x1b[39m\\x1b[49m \\x1b[33m\\x1b[1m test2 >\\x1b[22m\\x1b[39m \\x1b[33mwarning\\x1b[39m",
  ],
  [
    "\\x1b[90m01/01/1970 24:00:20.000 UTC\\x1b[39m \\x1b[48;5;15m\\x1b[38;5;15m[\\x1b[39m\\x1b[30mINFO   \\x1b[39m\\x1b[38;5;15m]\\x1b[39m\\x1b[49m \\x1b[1m test2 >\\x1b[22m info",
  ],
]
`;

snapshot[`log: custom config 1`] = `
[
  [
    "\\x1b[90m01/01/1970 24:00:02.000 UTC\\x1b[39m \\x1b[48;5;9m\\x1b[38;5;9m[\\x1b[39m\\x1b[30mERROR  \\x1b[39m\\x1b[38;5;9m]\\x1b[39m\\x1b[49m \\x1b[31m\\x1b[1m test1 >\\x1b[22m\\x1b[39m \\x1b[91merror\\x1b[39m",
  ],
  [
    "\\x1b[90m01/01/1970 24:00:04.000 UTC\\x1b[39m \\x1b[48;5;9m\\x1b[38;5;9m[\\x1b[39m\\x1b[30mERROR  \\x1b[39m\\x1b[38;5;9m]\\x1b[39m\\x1b[49m \\x1b[31m\\x1b[1m test1 >\\x1b[22m\\x1b[39m \\x1b[91mError : true error default
" +
      "
" +
      "[mocked error stack]\\x1b[39m",
  ],
  [
    "\\x1b[90m01/01/1970 24:00:16.000 UTC\\x1b[39m \\x1b[48;5;9m\\x1b[38;5;9m[\\x1b[39m\\x1b[30mERROR  \\x1b[39m\\x1b[38;5;9m]\\x1b[39m\\x1b[49m \\x1b[31m\\x1b[1m test2 >\\x1b[22m\\x1b[39m \\x1b[91merror\\x1b[39m",
  ],
  [
    "\\x1b[90m01/01/1970 24:00:18.000 UTC\\x1b[39m \\x1b[48;5;11m\\x1b[38;5;11m[\\x1b[39m\\x1b[30mWARNING\\x1b[39m\\x1b[38;5;11m]\\x1b[39m\\x1b[49m \\x1b[33m\\x1b[1m test2 >\\x1b[22m\\x1b[39m \\x1b[33mwarning\\x1b[39m",
  ],
  [
    "\\x1b[90m01/01/1970 24:00:20.000 UTC\\x1b[39m \\x1b[48;5;15m\\x1b[38;5;15m[\\x1b[39m\\x1b[30mINFO   \\x1b[39m\\x1b[38;5;15m]\\x1b[39m\\x1b[49m \\x1b[1m test2 >\\x1b[22m info",
  ],
  [
    "\\x1b[90m01/01/1970 24:00:22.000 UTC\\x1b[39m \\x1b[48;5;8m\\x1b[38;5;8m[\\x1b[39m\\x1b[30mDEBUG  \\x1b[39m\\x1b[38;5;8m]\\x1b[39m\\x1b[49m \\x1b[90m\\x1b[1m test2 >\\x1b[22m\\x1b[39m \\x1b[90mdebug\\x1b[39m",
  ],
  [
    "\\x1b[90m01/01/1970 24:00:24.000 UTC\\x1b[39m \\x1b[48;5;8m\\x1b[38;5;8m[\\x1b[39m\\x1b[30mVERBOSE\\x1b[39m\\x1b[38;5;8m]\\x1b[39m\\x1b[49m \\x1b[90m\\x1b[1m test2 >\\x1b[22m\\x1b[39m \\x1b[90mverbose\\x1b[39m",
  ],
]
`;
